raft.go

```diff
@@ -1375,7 +1376,23 @@ func (r *Raft) appendEntries(rpc RPC, a *AppendEntriesRequest) {
 		rpc.Respond(resp, rpcErr)
 	}()
 
-	// Ignore an older term
+	// this use case would happen when a node was a voter and is added back to the cluster as non voter,
+	// it term could be higher then the cluster, but because it's a non voter that term need to be discarded
+	// in favor of the cluster current term to keep the cluster stable, as an election don't need to be started
+	// by a node which don't have voting rights.
+	currentTerm := r.getCurrentTerm()
+	hasVote := hasVote(r.getLatestConfiguration(), r.localID)
+	if a.Term < currentTerm && !hasVote && a.PrevLogTerm >= r.lastLogTerm {
+		r.logger.Warn("older term sent to non-voter", "PrevLogTerm", a.PrevLogTerm,
+			"lastLogTerm", r.lastLogTerm,
+			" Term", a.Term, "currentTerm", currentTerm)
+		r.setState(Follower)
+		r.setCurrentTerm(a.Term)
```

Comment:
Is there any other state we need to set here 🤔 I don't _think_ so because this is the same as the code below for normal happy-path replication where we switch to a new leader's term after an election when we get the first `appendEntries` but I feel like we also store state about which node is our current leader in raft and it feels weird to me that we might be potentially switching term and leader here without updating anything about who that leader is....

But i know we do persist the identity of the candidate we last voted for, so I wonder if the happy path doesn't update anything because it assumes that we set the identity during election? Even then that can't be right in the case of a follower coming up and learning about a new leader that it didn't vote for so again I _guess_ you did the right thing here but I'm slightly intrigued about how and where we update the state about _who_ our current leader 🤔. Do you know @dhiaayachi ?

Reply:
it's done right after in [here](https://github.com/hashicorp/raft/blob/dbd6d068edfd1dda4ea4b0d960726165f9163bb8/raft.go#L1411-L1414)
but I'm a bit confused about it as that mean we trust any node who sends us an appendEntries request with a higher term as the leader, but maybe that's the right thing to do, I'm not sure.